src/pages/AddMembersPage/NewMemberList.tsx

```diff
@@ -0,0 +1,269 @@
+import React, { useRef, useState } from 'react';
+
+import { zodResolver } from '@hookform/resolvers/zod';
+import { useFieldArray, useForm } from 'react-hook-form';
+import { useQueryClient } from 'react-query';
+import { z } from 'zod';
+
+import { LoadingModal } from '../../components';
+import { isFeatureEnabled } from '../../config/features';
+import { zEthAddress } from '../../forms/formHelpers';
+import { useApeSnackbar, useApiBase } from '../../hooks';
+import { Check /* Working */ } from '../../icons/__generated';
+import { client } from '../../lib/gql/client';
+import { Box, Button, Flex, Panel, Text } from '../../ui';
+import { normalizeError } from '../../utils/reporting';
+
+import CopyCodeTextField from './CopyCodeTextField';
+import NewMemberEntry from './NewMemberEntry';
+import NewMemberGridBox from './NewMemberGridBox';
+
+const NewMemberList = ({
+  // TODO: figure out what to do w/ revoke
+  // revokeWelcome,
+  circleId,
+  welcomeLink,
+}: {
+  circleId: number;
+  welcomeLink: string;
+  revokeWelcome(): void;
+}) => {
+  const { fetchCircle } = useApiBase();
+  const { showError } = useApeSnackbar();
+
+  const [loading, setLoading] = useState<boolean>();
+  const [successCount, setSuccessCount] = useState<number>(0);
+
+  const successRef = useRef<HTMLDivElement>(null);
+
+  const queryClient = useQueryClient();
+
+  const newMemberSchema = z.object({
+    newMembers: z
+      .array(
+        z
+          .object({
+            address: zEthAddress.or(z.literal('')),
+            name: z
+              .string()
+              .min(3, 'Name must be at least 3 characters')
+              .or(z.literal('')),
+          })
+          .superRefine((data, ctx) => {
+            if (data.name && data.name !== '' && data.address === '') {
+              ctx.addIssue({
+                code: z.ZodIssueCode.custom,
+                path: ['address'],
+                message: 'Address is required if name is provided',
+              });
+            }
+            if (data.address && data.address !== '' && data.name === '') {
+              ctx.addIssue({
+                code: z.ZodIssueCode.custom,
+                path: ['name'],
+                message: 'Name is required if address is provided',
+              });
+            }
+          })
+      )
+      .superRefine(data => {
+        if (data.filter(m => m.address != '' && m.name != '').length == 0) {
+          //TODO: I want to use this to prevent form submission of there are no valid entries
+          // but it just breaks the more useful errors from rendering
+          // ctx.addIssue({
+          //   code: z.ZodIssueCode.custom,
+          //   message: 'no valid members entered',
+          // });
```

Comment:
I'm honestly not a huge fan of the the zodResolver myself. I think its a bit clumsy of an integration with react-hook-form. In cases like this where you want to prevent submission i create a clause in the submit button's `disabled` attribute.

Reply:
ok i used a watch and moved to submit button, good call